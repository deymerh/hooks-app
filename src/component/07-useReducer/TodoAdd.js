import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, resetValues] = useForm({
    description: ''
  });
  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 1) {
      return;
    };
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };
    handleAddTodo(newTodo);
    resetValues();
  };
  return (
    <>
      <h5>Agregar TODO</h5>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholder="Aprender..."
          className="form-control"
          autoComplete="off"
          onChange={handleInputChange}
          value={description}
        />
        <button
          type="submit"
          className="btn btn-outline-primary btn-block mt-2"
        >
          Agregar
          </button>
      </form>
    </>
  );
};
