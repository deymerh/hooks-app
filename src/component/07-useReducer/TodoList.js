import React from 'react';
import { TodoListItem } from './TodoListItem';
import propTypes from 'prop-types';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul className="list-group list-group-flush">
      {
        todos.map((todo, i) => (
          <TodoListItem
            key={todo.id}
            i={i}
            todo={todo}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        ))
      }
    </ul>
  );
};
TodoList.propTypes = {
  todos: propTypes.array.isRequired,
  handleDelete: propTypes.func.isRequired,
  handleToggle: propTypes.func.isRequired
};