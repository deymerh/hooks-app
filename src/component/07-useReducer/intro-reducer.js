const initialState = [{
  id: 1,
  task: 'Buy bread',
  done: false
}];

const taskReducer = (state = initialState, action) => {
  if (action?.type === 'AddTask') {
    return [...state, action.payload]
  }
  return state
};

let tasks = taskReducer();

//Nuevo objeto a agregar
const newTask = {
  id: 2,
  task: 'Run',
  done: false
}
//Accion a ejecutar
const AddTaskReducer = {
  type: 'AddTask',
  payload: newTask
}

tasks = taskReducer(tasks, AddTaskReducer)
console.log(tasks);