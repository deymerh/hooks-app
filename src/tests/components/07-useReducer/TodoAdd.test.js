import { shallow } from "enzyme";
import { TodoAdd } from "../../../component/07-useReducer/TodoAdd";

describe('Pruebas en componente <TodoAdd />', () => {
  const handleAddTodo = jest.fn();

  const wrapper = shallow(<TodoAdd
    handleAddTodo={handleAddTodo}
  />);

  test('Renderizar el componente correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('NO debe llamar el handleAddTodo', () => {
    const formSummit = wrapper.find('form').prop('onSubmit');
    formSummit({ preventDefault() { } })
    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('Debe llamar a la function handleAddTodo', () => {
    const value = 'Aprender React';
    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description'
      }
    });
    const formSummit = wrapper.find('form').prop('onSubmit');
    formSummit({ preventDefault() { } })
    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false
    });
    expect(wrapper.find('input').prop('value')).toBe('');
  });

});
