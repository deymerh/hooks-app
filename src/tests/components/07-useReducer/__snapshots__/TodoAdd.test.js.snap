// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <TodoAdd /> Renderizar el componente correctamente 1`] = `
<Fragment>
  <h5>
    Agregar TODO
  </h5>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-block mt-2"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
